### DISCUSSION TITLE
How to model receiving N consecutive `HEARTBEAT_MISSED` events before transitioning to 'OFFLINE' state?

### DISCUSSION THREAD

**Discussion by kad3nce** [statelyai/xstate#2016] [Q&A] [answered]
I only want my network monitor state machine to transition from `ONLINE` to `OFFLINE` if it receives N consecutive `HEARTBEAT_MISSED` events. If, for example, it receives one `HEARTBEAT_MISSED` event, followed by a `MESSAGE_RECEIVED` event, followed by another `HEARTBEAT_MISSED` event, I want it to stay in the `ONLINE` state. But say I have N configured to 2, if the state machine is `ONLINE` and receives two `HEARTBEAT_MISSED` events in a row, I want it to transition to `OFFLINE`.

What is the recommended way to model this behavior?

**Comment by Andarist** [comment, marked as answer]
1. Keep a counter in context
2. Increment it when receiving `HEARBEAT_MISSED`
3. Reset when you receive `MESSAGE_RECEIVED`
4. Use guarded transitions to transition to `OFFLINE` only when the counter has reached your threshold

**Reply by kad3nce** [reply]
Perfect. Thank you!
